[Zi Xuan] iP

zxgoh opened the pull request:
=============================================================================================



![JUKE](https://user-images.githubusercontent.com/73661051/152280664-189c867f-374f-4edb-99fa-cec2ed008df0.png)
> # Juke is ~cool~ awesome pawsome :dog::paw_prints:
Juke, an interactive chatbot, allows you freedom to plan and organize your tasks, such as:
-  _todo_
- _deadline_
- _event_

Downloading it is as easy as 1, 2, 3:
1. click this [link](https://github.com/zxgoh/ip/releases/tag/A-Jar)
2. download ip.jar
3. you're done, voilà :disguised_face:

**<div align="center">Here's a sneak peek into the inner-workings of Juke, using [filewriter](https://docs.oracle.com/javase/7/docs/api/java/io/FileWriter.html)</div>**
```
private static void writeToFile(String filePath, String textToAdd) throws IOException {
    FileWriter fw = new FileWriter(filePath, true);
    fw.write(textToAdd);
    fw.close();
    }
```
=============================================================================================

Commit:
Add Gradle support

zxgoh committed:
Implement a skeletal version of Duke that starts by greeting the user, simply echos commands entered by the user, and exits when the user types bye.

zxgoh committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested.

zxgoh committed:
Add the ability to mark tasks as done. Optionally, add the ability to change the status back to not done.

zxgoh committed:
to ignore

zxgoh committed:
Add support for tracking three types of tasks

zxgoh committed:
Add support for deleting tasks from the list.

zxgoh committed:
Teach Duke to deal with errors such as incorrect inputs entered by the user.

zxgoh committed:
Added comments to the various if-else checks to describe functionality

zxgoh committed:
included the expected outcome when running Duke.java

zxgoh committed:
Save task automatically in hard disk

zxgoh committed:
Save task automatically in hard disk

zxgoh committed:
Convert numerical date and time to understandable sentence-like date and time

zxgoh committed:
Merge branch 'branch-Level-8'

zxgoh committed:
add .gitignore for duke.txt

zxgoh committed:
no message

zxgoh committed:
Abstract the relevant methods into categorised classes, MoreOOP

zxgoh committed:
Package the relevant classes into package named "duke"

zxgoh committed:
Add JUnit tests to test the behavior of the code

zxgoh committed:
Create Jar file for testing usability of the interface and application

zxgoh committed:
Made changes to Level 7

zxgoh committed:
Add javadocs header to classes and method

zxgoh committed:
Ensure coding standard through proper indentation, layout, and javadocs

zxgoh committed:
Add search feature to find tasks matching the relevant keyword

zxgoh committed:
Merge branch 'A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Todo.java

zxgoh committed:
Merge branch 'branch-Level-9'

zzkzzzz commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,173 @@
+    public static void main(String[] args) {
+
+        Outputs op = new Outputs();
+
+
+
```
Should these empty lines be deleted?

zzkzzzz commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,173 @@
+                } else { //valid todo command
+                    Task todoTask = new Todo(reply);
+                    itemList.add(todoTask);
+                    System.out.println(op.border +
+                            "     Got it. I've added this task: \n" +
+                            "       " + todoTask.getDescription() + "\n" +
+                            "     Now you have " + itemList.size() + " tasks in the list.\n" +
+                            op.instructions +
+                            op.border);
```
I like how you separated the lines which improve readability.

AryanSarswat replied:
You could consider using StringBuilder from java, it improves the speed of your program as you know concatenating string in java is very slow as they have to create a new String and add all the chars to the new string one by one. StringBuilder simpler does what you expect where it does not need to create a new String, it just "appends" the new string at the end!

This is optional tho, using String concatenation is completely fine aswell!

zzkzzzz commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,80 @@
+     */
+    public String getDescription() {
+
+        String newReply = super.description.replace("deadline ", "");
+        String taskAtHand = newReply.split("/")[0];
+        String deadLine = newReply.split("/")[1].replace("by ", "by: ");
+        String date = deadLine.split(" ")[1].split("-")[0];
+        String month = digitMonth_to_AlphabeticalMonth(deadLine.split(" ")[1].split("-")[1]);
+        String year = deadLine.split(" ")[1].split("-")[2];
+        String time = Integer.toString(Integer.parseInt(deadLine.split(" ")[2]) % 1200);
```
I like how you name these variables which follow the code standard.

zxgoh replied:
Thank you @zzkzzzz  for the kind feedback, will make the changes in the next few commits 🥂

zzkzzzz commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,80 @@
+     *
+     * @param digitMonth the month represented numerically
+     * @return corresponding month in words if valid, else returns notValidMonth
+     * @throws
+     */
+    public String digitMonth_to_AlphabeticalMonth(String digitMonth) {
```
I think the methods name should use camelCase instead of underscores.

zzkzzzz commented on src/main/java/duke/Outputs.java:
```diff
@@ -0,0 +1,39 @@
+
+    }
+
+    String greeting = "Welcome, fellow user, to\n";
+    String logo =         "\n" +
+            "       _       _        \n" +
+            "      | |     | |       \n" +
+            "      | |_   _| | _____ \n" +
+            "  _   | | | | | |/ / _ \\\n" +
+            " | |__| | |_| |   <  __/\n" +
+            "  \\____/ \\__,_|_|\\_\\___|\n" +
+            "                        \n" +
+            "                        \n";
+
+    String instructions =
+            "\n" +
+                    "     **********************************************************************\n" +
+                    "     * COMMANDS | FORMAT                                                  *\n" +
+                    "     * normal task |  todo [taskname]                                     *\n" +
+                    "     * normal task with deadline | deadline [taskname] /by [date&time]    *\n" +
+                    "     * event task with deadline | event [taskname] /by [date&time]        *\n" +
+                    "     *                                                                    *\n" +
+                    "     * Note for [date&time]: enter it strictly in the following format    *\n" +
+                    "     *    dd/mm/yyyy [24h] | eg: 31-12-1969 1830                          *\n" +
+                    "     **********************************************************************\n" +
+                    "\n";
+
+    String firstPrompt =                 "     Hello! I'm Juke\n" +
+            "     What can I do for you?\n";
+
+    String border = "    ‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾\n";
```
Should these constant use uppercase?

zzkzzzz reviewed (commented):
Overall, I found your code is very clear and easy to read except in a few places
where the naming of variables was inappropriate. And I noted a few minor coding standard violations too. Hope it helps!  Thank you!

zxgoh committed:
Automate project builds using Gradle

zxgoh committed:
Merge branch 'add-gradle-support'

zxgoh commented:
> Overall, I found your code is very clear and easy to read except in a few places where the naming of variables was inappropriate. And I noted a few minor coding standard violations too. Hope it helps! Thank you!

Thank you for the kind feedback, will take note of the issues and fix them in the next few commits 🥂

AryanSarswat commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,80 @@
+package duke;
+/**
+ * Represents the deadline tasks input by the user
```
Consider ending your descriptions with a period. (This is a general comment and applies to all your JavaDoc comments)

AryanSarswat commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,80 @@
+package duke;
+/**
```
Consider adding a separation line between package duke; and your java doc comment.

Generally speaking, we should add spaces between methods, between imports and the class etc.

This may be relevant for other parts of your code as well, it helps improve the look of your code!

AryanSarswat commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,80 @@
+        } else if (digitMonth.equals("12")) {
+            return "December";
+        } else {
+            return "notValidMonth";
```
You could maybe consider using the LocalDateTime library, which handles all the month, year, and day calculations for you!

AryanSarswat commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,80 @@
+     * @param
+     * @return deadline task in a string format, changing the numerical months to
+     * words, and 24h time to 12h time
+     * @throws
```
Do you need the throws javadoc comment here?

AryanSarswat commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,80 @@
+        String month = digitMonth_to_AlphabeticalMonth(deadLine.split(" ")[1].split("-")[1]);
+        String year = deadLine.split(" ")[1].split("-")[2];
+        String time = Integer.toString(Integer.parseInt(deadLine.split(" ")[2]) % 1200);
+        if (time.length() < 4 ) {
```
Consider removing the whitespace after 4.

AryanSarswat commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,173 @@
+                file.createNewFile();
+            } catch (IOException e) {
+                System.out.println(" ");
+            }
```
You should consider making a Storage.java class to handle the creation of the storage folder and file.
This applies to lines 163-169 aswell.

AryanSarswat commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,78 @@
+     * Super constructor to the task class
+     *
+     * @param description the description of the event task
+     * @return
```
Maybe remove the @return parameter from your javadoc as the Constructor class does not actually return anything?

AryanSarswat commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,78 @@
+        } else if (digitMonth.equals("12")) {
+            return "December";
+        } else {
+            return "notValidMonth";
```
Maybe consider abstracting this method out into the Task,java class making it static. Thus when you inherit this method in Event and Deadline you do not have to rewrite this code again.

This may also apply to other methods like getDescription(). Where you could have a method which returns the string in the following format:
```java
return this.taskName + this.description
``` 
(You may need to create an attribute called this.taskName for this)
Afterwhich you can overwrite it in deadline and event classes to add on the dates like so

```java
return super.getDescription() + this.date
```

This helps with reducing your workload!

AryanSarswat commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,56 @@
+     * @throws
+     */
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
```
Nice usage of the ternary operator! 😃

AryanSarswat commented on src/main/java/duke/Todo.java:
```diff
@@ -0,0 +1,30 @@
+     * @return todo task in a string format
+     * @throws
+     */
+    public String getDescription() {
+        String newReply = super.description.replace("todo ", "");
+        return "[T]" + "[" + super.getStatusIcon() + "] " + newReply;
+    }
```
You should also consider overwriting the toString() method of the task instead of creating a sepearte getDescription() method as the toString() method is not being used for anything else so you might as well repurpose it. This is will also help with printing as you do:
```Java 
System.out.println(event) 
``` 
instead of printing out the hashcode for the object its description will appear.

AryanSarswat commented on text-ui-test/EXPECTED.TXT:
```diff
@@ -1,7 +1,15 @@
-| |_| | |_| |   <  __/
-|____/ \__,_|_|\_\___|
 
+       _       _
+      | |     | |
+      | |_   _| | _____
+  _   | | | | | |/ / _ \
+ | |__| | |_| |   <  __/
+  \____/ \__,_|_|\_\___|
+
+
+
+    ____________________________________________________________
+     Hello! I'm Juke
+     What can I do for you?
+    ____________________________________________________________
```
I like the name of your Bot 👍

AryanSarswat reviewed (commented):
Overall a great job done, other than a few coding standards changes you are good to go!
Hope some of my comments regarding the abstraction of some methods were useful and help reduce the duplication of some of your code!

zxgoh committed:
Add a GUI to Duke. Use the JavaFX technology to implement the GUI.

zxgoh committed:
Add a GUI to Duke. Use the JavaFX technology to implement the GUI.

zxgoh committed:
Add minor changes to code

zxgoh committed:
Update README.md

zxgoh committed:
Update README.md

zxgoh committed:
Remove the previously used Main() method

Fix the bug where application will now close upon entering a "bye"
command

Change the filepath for mainClassName in build.gradle to
duke.Launcher

zxgoh committed:
Merge branch 'master' of https://github.com/zxgoh/ip

zxgoh committed:
Use assert feature (not JUnit assertions) to document important assumptions that should hold at various points in the code.

zxgoh committed:
Improve code quality

zxgoh committed:
Merge pull request #2 from zxgoh/branch-A-CodeQuality

Improve code quality

zxgoh committed:
Merge branch 'master' into branch-A-Assertions

zxgoh committed:
Add tagging function so user can categorise tasks

zxgoh committed:
made minor changes

zxgoh committed:
Merge branch 'branch-A-Assertions'

# Conflicts:
#	src/main/java/duke/Duke.java

zxgoh committed:
Improve the GUI

zxgoh committed:
Add Ui.png

zxgoh committed:
Update README.md

Friendly user guide

zxgoh committed:
Change specifications on build.gradle

zxgoh committed:
Merge remote-tracking branch 'origin/master'

zxgoh committed:
Update README.md

zxgoh committed:
Set theme jekyll-theme-minimal

zxgoh committed:
Update README.md

zxgoh committed:
Fix minor date and month bugs

zxgoh committed:
Fix build.gradle file

zxgoh committed:
Update README.md

zxgoh committed:
Fix bugs

zxgoh committed:
fix bugs

zxgoh committed:
Implement necessary JUnit tests

zxgoh committed:
Update JavaDocs

zxgoh committed:
Update JavaDocs

zxgoh committed:
Update help command output

zxgoh committed:
Update help command outputs

zxgoh committed:
Add commands to UserGuide

zxgoh committed:
Make minor changes to code

zxgoh committed:
Update README.md
